python/visualization.py

```diff
@@ -988,27 +989,164 @@ def plot2D(
     return ax
 
 
-def plot3D(sim: Simulation):
-    from mayavi import mlab
+def plot3D(sim):
+    from vispy.scene.visuals import Box, Mesh
+    from vispy.scene import SceneCanvas, transforms
 
-    if sim.dimensions < 3:
-        raise ValueError("Simulation must have 3 dimensions to visualize 3D")
+    try:
+        from skimage.measure import marching_cubes
+    except:
+        from skimage.measure import marching_cubes_lewiner as marching_cubes
+    from vispy.visuals.filters import ShadingFilter
 
-    xmin, xmax, ymin, ymax, zmin, zmax = box_vertices(
-        sim.geometry_center, sim.cell_size
+    # Set canvas
+    canvas = SceneCanvas(keys="interactive", bgcolor="white")
+
+    view = canvas.central_widget.add_view()
+    view.camera = "arcball"
+
+    # Get domain measurements
+    sim_center, sim_size = sim.geometry_center, sim.cell_size
+
+    xmin, xmax, ymin, ymax, zmin, zmax = mp.visualization.box_vertices(
+        sim_center, sim_size, sim.is_cylindrical
     )
 
-    Nx = int(sim.cell_size.x * sim.resolution) + 1
-    Ny = int(sim.cell_size.y * sim.resolution) + 1
-    Nz = int(sim.cell_size.z * sim.resolution) + 1
+    grid_resolution = sim.resolution
+
+    Nx = int((xmax - xmin) * grid_resolution + 1)
+    Ny = int((ymax - ymin) * grid_resolution + 1)
+    Nz = int((zmax - zmin) * grid_resolution + 1)
 
     xtics = np.linspace(xmin, xmax, Nx)
     ytics = np.linspace(ymin, ymax, Ny)
     ztics = np.linspace(zmin, zmax, Nz)
 
-    eps_data = sim.get_epsilon_grid(xtics, ytics, ztics)
-    s = mlab.contour3d(eps_data, colormap="YlGnBu")
-    return s
+    # Get eps for geometry
+    eps_data = np.round(np.real(sim.get_epsilon_grid(xtics, ytics, ztics)), 2)
+
+    unique = np.unique(np.abs(eps_data)).tolist()
+
+    # Remove background material
+    unique.remove(np.round(np.abs(np.asarray(sim.default_material.epsilon_diag)), 2)[0])
+
+    mesh_midpoint = (sim_size[0] / 2, sim_size[1] / 2, sim_size[2] / 2)
+
+    # Build geometry
+    for i, eps in enumerate(unique):
+        eps_ = np.array(eps_data.flatten() == eps).astype(int).reshape(eps_data.shape)
+        marching_cube = marching_cubes(
+            eps_,
+            0.99,
+            spacing=(sim.cell_size.x / Nx, sim.cell_size.y / Ny, sim.cell_size.z / Nz),
+        )
+        vertices, faces = marching_cube[0], marching_cube[1]
+
+        mesh = Mesh(
+            vertices,
+            faces,
+            color=(
```

Comment:
It might be good to choose a default color for the mesh which has a lighter tone than the image shown above (similar to the original image from mayavi) in order to be able to better visualize the contrast in the surfaces when rotating the object around given a fixed angle for the illumination.

Reply:
Hello @oskooi,

Sorry for the late response.

vispy can render the scene as a numpy array which can be saved as an image. See here:

![sim](https://user-images.githubusercontent.com/100642027/200966992-f1771387-f785-4223-aafe-e633f168baeb.png)

As for the image quality, we can add kwargs to the function to define the camera angles (in cylindrical coordinates). The shading is affected by the camera angles.

What do you think?

Reply:
> vispy can render the scene as a numpy array which can be saved as an image.

I suppose you are referring to the function [`SceneCanvas.render`](https://vispy.org/api/vispy.scene.canvas.html#vispy.scene.canvas.SceneCanvas.render)? This is what we should use in the mini example.

> As for the image quality, we can add kwargs to the function to define the camera angles (in cylindrical coordinates).

Adding kwargs is the right approach as it is consistent with `plot2D`.

Reply:
> > vispy can render the scene as a numpy array which can be saved as an image.
> 
> I suppose you are referring to the function [`SceneCanvas.render`](https://vispy.org/api/vispy.scene.canvas.html#vispy.scene.canvas.SceneCanvas.render)? This is what we should use in the mini example.

Agreed, that is what I used to generate the image in the above comment.

> > As for the image quality, we can add kwargs to the function to define the camera angles (in cylindrical coordinates).
> 
> Adding kwargs is the right approach as it is consistent with `plot2D`.

Sweet, i will add these to this PR.